src/backend/distributed/utils/node_metadata.c

```diff
@@ -264,13 +274,150 @@ master_remove_node(PG_FUNCTION_ARGS)
+	{
+		ereport(ERROR, (errmsg(
+							"cannot mark a node for draining that is not a currently a data node")));
+	}
+	SetIsDataNode(workerNode, IsDataNodeMarkedForRemovalId());
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * master_make_nodata_node sets the isdatanode field in pg_dist_node to
+ * 'false'. This fails when there are primary placements on this node, these
+ * should be moved to other nodes first.
+ */
+Datum
+master_make_nodata_node(PG_FUNCTION_ARGS)
+{
+	text *nodeNameText = PG_GETARG_TEXT_P(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	WorkerNode *workerNode = ModifiableWorkerNodeWithoutPrimaryPlacements(nodeNameText,
+																		  nodePort);
+	SetIsDataNode(workerNode, IsDataNodeFalseId());
+	CleanUpReferenceTables(workerNode);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * master_make_data_node sets the isdatanode field in pg_dist_node to 'true',
+ * i.e new distributed tables will get placements on this node.
+ */
+Datum
+master_make_data_node(PG_FUNCTION_ARGS)
+{
+	text *nodeNameText = PG_GETARG_TEXT_P(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	WorkerNode *workerNode = ModifiableWorkerNode(nodeNameText, nodePort);
+	WorkerNode *newWorkerNode = SetIsDataNode(workerNode, IsDataNodeTrueId());
+
+	SetUpReferenceTablesAndDistributedFunctions(newWorkerNode);
+
+	PG_RETURN_VOID();
+}
+
 
+/*
+ * CleanupreferenceTables updates the replication factor for reference tables
+ * to be one lower if the node had reference tables before and doesn't now.
+ */
+static void
+CleanUpReferenceTables(WorkerNode *workerNode)
+{
+	if (WorkerNodeIsPrimaryCurrentDataNode(workerNode))
+	{
+		UpdateColocationGroupReplicationFactorForReferenceTables(
+			list_length(ActivePrimaryCurrentDataNodeList(NoLock)));
+	}
+}
+
+
+/*
+ * SetUpReferenceTablesAndDistributedFunctions sets up reference tables and
+ * distributed functions on a node if it should store distributed data now.
+ */
+static void
+SetUpReferenceTablesAndDistributedFunctions(WorkerNode *newWorkerNode)
+{
+	if (WorkerNodeIsPrimaryCurrentDataNode(newWorkerNode))
+	{
+		EnsureNoModificationsHaveBeenDone();
+		ReplicateAllDependenciesToNode(newWorkerNode->workerName,
+									   newWorkerNode->workerPort);
+		ReplicateAllReferenceTablesToNode(newWorkerNode->workerName,
+										  newWorkerNode->workerPort);
+
+		/*
+		 * Let the maintanince deamon do the hard work of syncing the metadata. We prefer
+		 * this because otherwise node activation might fail withing transaction blocks.
+		 */
+		if (ClusterHasDistributedFunctionWithDistArgument())
+		{
+			MarkNodeHasMetadata(newWorkerNode->workerName, newWorkerNode->workerPort,
+								true);
+			TriggerMetadataSync(MyDatabaseId);
+		}
+	}
+}
+
+
+/*
+ * ModifiableWorkerNode gets the requested WorkerNode and also gets locks
+ * required for modifying it. This fails if the node does not exist.
+ */
+static WorkerNode *
+ModifiableWorkerNode(text *nodeNameText, int32 nodePort)
+{
+	char *nodeName = text_to_cstring(nodeNameText);
```

Comment:
as said earlier, use `char *`'s to pass the name instead of this late postgres to c translation.